blivet/nvdimm.py

```diff
@@ -0,0 +1,181 @@
+program_log = logging.getLogger("program")
+
+
+log_bd_message = lambda level, msg: program_log.info(msg)
+
+
+_REQUESTED_PLUGINS = BlockDev.plugin_specs_from_names(("nvdimm",))
+_HAVE_NVDIMM = False
+
+try:
+    # do not check for dependencies during libblockdev initializtion, do runtime
+    # checks instead
+    BlockDev.switch_init_checks(False)
+    succ_, avail_plugs = BlockDev.try_reinit(require_plugins=_REQUESTED_PLUGINS,
+                                             reload=False,
+                                             log_func=log_bd_message)
+except GLib.GError as err:
+    log.error("Failed to intialize the libblockdev library: %s", err)
+    _HAVE_NVDIMM = False
+else:
+    _HAVE_NVDIMM = "nvdimm" in avail_plugs
+
+
+class NVDIMMDependencyGuard(util.DependencyGuard):
+    error_msg = "libblockdev NVDIMM functionality not available"
+
+    def _check_avail(self):
+        if not _HAVE_NVDIMM:
+            return False
+
+        try:
+            BlockDev.nvdimm_is_tech_avail(BlockDev.NVDIMMTech.NVDIMM_TECH_NAMESPACE,
+                                          BlockDev.NVDIMMTechMode.RECONFIGURE |
+                                          BlockDev.NVDIMMTechMode.QUERY |
+                                          BlockDev.NVDIMMTechMode.ACTIVATE_DEACTIVATE)
+        except GLib.GError:
+            return False
+        return True
+
+blockdev_nvdimm_required = NVDIMMDependencyGuard()
+
+
+log_bd_message = lambda level, msg: program_log.info(msg)
+
+
+class NVDIMM(object):
+    """ NVDIMM utility class.
+
+        .. warning::
+            Since this is a singleton class, calling deepcopy() on the instance
+            just returns ``self`` with no copy being created.
+    """
+
+    nvdimm_plugin_available = False
+
+    def __init__(self):
+        self._namespaces = None
+
+    # So that users can write nvdimm() to get the singleton instance
+    def __call__(self):
+        return self
+
+    def __deepcopy__(self, memo_dict):
+        # pylint: disable=unused-argument
+        return self
+
+    @property
+    def namespaces(self):
+        """ Dict of all NVDIMM namespaces, including dax and disabled namespaces
+        """
+        if not self._namespaces:
+            self.update_namespaces_info()
+
+        return self._namespaces
+
+    @blockdev_nvdimm_required(critical=True, eval_mode=util.EvalMode.onetime)
+    def update_namespaces_info(self):
+        """ Update information about the namespaces
+        """
+        namespaces = BlockDev.nvdimm_list_namespaces(idle=True)
+
+        self._namespaces = dict((namespace.dev, namespace) for namespace in namespaces)
+
+    def get_namespace_info(self, device):
+        """ Get namespace information for a device
+            :param str device: device name (e.g. 'pmem0') or path
+        """
+        for info in self.namespaces.values():
+            if info.blockdev == device or \
+               (device.startswith("/dev/") and info.blockdev == device[5:]):
```

Comment:
You could simplify it slightly by doing `if info.blockdev == device.split("/")[-1]` or `if info.blockdev == os.path.basename(device)`. Maybe not worthwhile if this code is already on 3.1-devel.